src/bsp/src/posix/src/InterlocManager.cpp

```diff
@@ -0,0 +1,148 @@
+    return distantAgentTf.inverseTimes(currentAgentTf);
+}
+
+double InterlocManager::getDistance(const tf2::Transform& transform) {
+    return sqrt(pow(transform.getOrigin().x(), 2) + pow(transform.getOrigin().y(), 2) +
+                pow(transform.getOrigin().z(), 2));
+}
+
+double InterlocManager::getOrientation(const tf2::Transform& transform) {
+    double roll;
+    double pitch;
+    double yaw;
+    tf2::Matrix3x3(transform.getRotation()).getRPY(roll, pitch, yaw);
+
+    return yaw;
+}
+
+std::optional<geometry_msgs::TransformStamped> InterlocManager::getHiveBoardTransform(
+    const std::string& agentName) {
+    geometry_msgs::TransformStamped transform;
+
+    try {
+        transform = m_tfBuffer.lookupTransform(agentName + HIVEBOARD_SUFFIX,
+                                               agentName + ROBOT_CENTER_SUFFIX, ros::Time(0));
+    } catch (tf2::TransformException& ex) {
+        m_logger.log(LogLevel::Warn, "Could not find transform to hiveboard for %s",
+                     agentName.c_str());
+        m_logger.log(LogLevel::Warn, "%s", ex.what());
+    }
+
+    return transform;
+}
+
+void InterlocManager::gazeboUpdateCallback(const gazebo_msgs::ModelStates& msg) {
+    const uint16_t currentAgentId = BSPContainer::getBSP().getUUId();
+    std::map<uint16_t, uint16_t> idxLookup;
+    const std::regex agentIdRegex("agent_([0-9]+)");
+    std::smatch match;
+
+    // Prepare map of id <-> index
+    for (unsigned int i = 0; i < msg.name.size(); i++) {
+        if (std::regex_match(msg.name[i], match, agentIdRegex)) {
+            uint16_t agentId = stoi(match.str(1));
+
+            // Only try to get the transform once per robot as it will not change at runtime
+            if (m_baseLinkToHiveBoardTransforms.count(agentId) == 0) {
+                auto transform = getHiveBoardTransform(msg.name[i]);
+                if (!transform) {
+                    // Transform between robot and hiveboard could not be found. Do not add this ID
+                    // to the map to prevent calculating interloc with it.
+                    continue;
+                }
+                m_baseLinkToHiveBoardTransforms[agentId] = transform.value();
+            }
+
+            idxLookup[i] = agentId;
+        }
+    }
+
+    if (idxLookup.count(currentAgentId) == 0) {
+        return;
+    }
+
+    for (auto& x : idxLookup) {
+        uint16_t index = x.first;
+        uint16_t agentId = x.second;
+
+        if (agentId == currentAgentId) {
+            // Do not calculate interloc with ourself
+            continue;
+        }
+
+        tf2::Transform relTransform = computeRelativeTransform(
+            msg.pose[idxLookup[currentAgentId]], m_baseLinkToHiveBoardTransforms[currentAgentId],
+            msg.pose[index], m_baseLinkToHiveBoardTransforms[agentId]);
+
+        double distance = getDistance(relTransform);
+        double orientation = getOrientation(relTransform) * 180 / M_PI;
```

Comment:
I was wondering, is there a reason why we are in degrees and not rad? Most sine and cosine functions uses rad. Is it since it's easier for precision? (i.e. more precise with fixed point)

Reply:
I was using degrees just because it is easier for us to understand, but we could switch everything to rads if you want to